### PR TITLE
Convert symmetric ciphers to Rust

### PR DESCRIPTION
**Pull request by alex**
TODO:

- [x] https://github.com/sfackler/rust-openssl/pull/2093
- [x] https://github.com/sfackler/rust-openssl/pull/2090
- [x] https://github.com/sfackler/rust-openssl/pull/2100
- [x] rust-openssl release
- [x] LibreSSL version bump
- [x] Chunking
- [x] Coverage
  - [x] `PyAEADEncryptionContext::update_into` bytes remaining check
  - [x] `PyAEADDecryptionContext::update` bytes remaining check
  - [x] `PyAEADDecryptionContext::update_into` bytes remaining check
  - [x] `PyAEADDecryptionContext::authenticate_additional_data` bytes remaining check
  - [x] `PyAEADDecryptionContext::finalize_with_tag` duplicate tag check
  - [x] `AlreadyUpdated` in `PyAEADDecryptionContext::authenticate_additional_data`
  - [x] 0 length `CffiMutBuf`
  - [x] XTS error checks

### CODE REVIEW

**Comment by jlaine** [comment]
Hi @alex! I'm hitting a deprecation regarding ARC4 when I use this branch, even when I'm not touching ARC4:

```python
>>> from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
>>> Cipher(algorithm=algorithms.AES(bytes(16)), mode=modes.ECB()).encryptor()
/path/to/env/lib/python3.11/site-packages/cryptography/hazmat/primitives/ciphers/base.py:107: CryptographyDeprecationWarning: ARC4 has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.ARC4 and will be removed from this module in 48.0.0.
  return rust_openssl.ciphers.create_encryption_ctx(
```

**Comment by alex** [comment]
I haven't updated this branch since @reaperhulk deprecated ARC4, I think `types.rs` just needs a simple update.

**Comment by reaperhulk** [comment]
https://github.com/secdev/scapy/pull/4285 should unblock us here.

File: src/cryptography/hazmat/bindings/_rust/openssl/ciphers.pyi [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+) -> ciphers.AEADDecryptionContext: ...
+@typing.overload
+def create_decryption_ctx(
+    algorithm: ciphers.CipherAlgorithm, mode: modes.Mode
+) -> ciphers.CipherContext: ...
+def cipher_supported(
+    algorithm: ciphers.CipherAlgorithm, mode: modes.Mode
+) -> bool: ...
+def _advance(
+    ctx: ciphers.AEADEncryptionContext | ciphers.AEADDecryptionContext, n: int
```
Comment (reaperhulk) [inline src/cryptography/hazmat/bindings/_rust/openssl/ciphers.pyi:30]: I think you need to use `typing.Union` here still. The `|` syntax is py310+

Comment (alex) [reply src/cryptography/hazmat/bindings/_rust/openssl/ciphers.pyi:30]: It's in a pyi file, which means it's parsed by mypy, not Python, which means I don't.

File: src/rust/src/backend/ciphers.rs [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+                ),
+            ));
+        }
+
+        let min_tag_length = ctx
+            .py_mode
+            .as_ref(py)
+            .getattr(pyo3::intern!(py, "_min_tag_length"))?
+            .extract()?;
+        // XXX: Do not hard code 16
```
Comment (reaperhulk) [inline src/rust/src/backend/ciphers.rs:447]: Stray comment

**Review by reaperhulk — APPROVED** [review]

**Pull request merged**